README.md
# Bloonix FastCGI wrapper

## About Bloonix

For general information about Bloonix, questions, ratings or other things please visit our base repository:

* https://github.com/bloonix/bloonix


ChangeLog
bloonix-fcgi (0.8) released at 2015-03-28

    * Extra release because the gpg key of bloonix is updated.

bloonix-fcgi (0.7) released at 2015-03-21

    * Moved SharedFile.pm and SharedMem.pm to the core package of Bloonix.
    * Moved process management to the core package or Bloonix.
    * Implemented a timeout funcionality to workaround blocking accept().

bloonix-fcgi (0.6) released at 2015-03-09

    * Suppress fatal messages for malformed json strings.

bloonix-fcgi (0.5) released at 2015-03-09

    * Improved output of server-status.

bloonix-fcgi (0.4) released at 2014-12-08

    * Ignoring "child $pid died 13".
    * New parameter max_program_size that defaults to 1GB.
    * USR1 and USR2 signals are now forwarded to all children.

bloonix-fcgi (0.3) released at 2014-11-03

    * Fixed: try to fixing malformed utf8 strings.
    * Updated the license information.

bloonix-fcgi (0.2) released at 2014-10-16

    * Improved logging for signals hup, int, term and pipe.

bloonix-fcgi (0.1) released at 2014-08-25

    * Initial release.


lib/Bloonix/FCGI.pm
package Bloonix::FCGI;

use strict;
use warnings;
use Bloonix::FCGI::Request;
use IO::Socket;
use FCGI;
use Params::Validate qw();
use POSIX qw(:sys_wait_h);
use Log::Handler;
use Time::HiRes qw();

use constant PARENT_PID => $$;

use base qw(Bloonix::Accessor);
__PACKAGE__->mk_accessors(qw/log proc request sock/);

our $VERSION = "0.8";

sub new {
    my $class = shift;
    my $opts = $class->validate(@_);
    my $self = bless $opts, $class;
    $self->init;
    return $self;
}

sub init {
    my $self = shift;

    $self->log(Log::Handler->get_logger("bloonix"));
    $self->init_fcgi_socket;
    $self->init_fcgi_request;
}

sub init_fcgi_socket {
    my $self = shift;

    $self->log->info("start socket on port :$self->{port}");
    $self->{sock} = FCGI::OpenSocket(
        ":".$self->{port},
        $self->{listen},
    );

    my $oldfh = select STDOUT;
    $| = 1;
    select STDERR;
    $| = 1;
    select $oldfh;
}

sub init_fcgi_request {
    my $self = shift;

    $self->{request} = FCGI::Request(
        \*STDIN,
        \*STDOUT,
        \*STDERR,
        \%ENV,
        $self->sock,
        &FCGI::FAIL_ACCEPT_ON_INTR
    );
}

sub accept {
    my ($self, $timeout) = @_;

    if ($timeout) {
        return $self->accept_timeout($timeout);
    }

    return $self->request->Accept() >= 0;
}

sub accept_timeout {
    my ($self, $timeout) = @_;
    my $ret;

    eval {
        local $SIG{__DIE__} = sub { alarm(0) };
        local $SIG{ALRM} = sub { die "FCGI ACCEPT TIMEOUT" };
        Time::HiRes::alarm($timeout);
        $ret = $self->request->Accept() >= 0;
        alarm(0);
    };

    return $ret;
}

sub get_new_cgi {
    my $self = shift;

    return Bloonix::FCGI::Request->new();
}

sub attach {
    my $self = shift;

    return $self->request->Attach();
}

sub detach {
    my $self = shift;

    return $self->request->Detach();
}

sub finish {
    my $self = shift;

    return $self->request->Finish();
}

sub validate {
    my $class = shift;

    my %options = Params::Validate::validate(@_, {
        port => {
            type => Params::Validate::SCALAR,
            default => 9000
        },
        listen => {
            type => Params::Validate::SCALAR,
            default => SOMAXCONN # from IO::Socket
        }
    });

    return \%options;
}

sub DESTROY {
    if ($$ == PARENT_PID) {
        FCGI::CloseSocket(shift->{sock});
    }
}

1;

=head1 NAME

Bloonix::FCGI - The database interface.

=head1 SYNOPSIS

=head1 DESCRIPTION

=head1 METHODS

=head1 EXPORTS

No exports.

=head1 REPORT BUGS

Please report all bugs to <support(at)bloonix.de>.

=head1 AUTHOR

Jonny Schulz <support(at)bloonix.de>.

=head1 COPYRIGHT

Copyright (C) 2009 by Jonny Schulz. All rights reserved.

=cut


lib/Bloonix/FCGI/Request.pm
package Bloonix::FCGI::Request;

use strict;
use warnings;
use Encode;
use JSON;
use MIME::Base64 ();

use base qw(Bloonix::Accessor);
__PACKAGE__->mk_accessors(qw/content_type content_types charset boundary postdata jsondata is_json maxlen log/);

use constant OS => $^O || do {
    no warnings "once";
    require Config;
    $Config::Config{"osname"};
};

use constant CRLF => do {
    OS =~ m/VMS/i
        ? "\n"
        : "\t" ne "\011"
            ? "\r\n"
            : "\015\012";
};

sub new {
    my ($class, %opts) = @_;

    my $self = bless {
        params => { },
        maxlen => $opts{maxlen} // 5242880,
        log => Log::Handler->get_logger("bloonix"),
    }, $class;

    my $content_type = $ENV{CONTENT_TYPE} || "";
    $content_type =~ s/\s//g;

    my @content_types = split /\s*;\s*/, $content_type;
    $self->{content_type} = shift @content_types || "";

    foreach my $pair (@content_types) {
        my ($name, $value) = split /=/, $pair, 2;

        if ($name eq "charset" || $name eq "boundary") {
            $self->{$name} = $value;
        }

        $self->{content_types}->{$name} = $value;
    }

    $self->{charset} ||= "";
    $self->{boundary} ||= "";
    $self->{is_json} = $self->content_type eq "application/json";
    $self->_process_post_data;
    $self->_process_query_string($self->query_string);

    return $self;
}

sub param {
    my ($self, $param) = @_;
    my $params = $self->{params};

    if (!defined $param) {
        die "no param set";
    }

    if (!exists $params->{$param}) {
        return wantarray ? () : undef;
    }

    if (ref $params->{$param} ne "ARRAY") {
        # Normally () is not necessary...
        return wantarray ? ($params->{$param}) : $params->{$param};
    }

    if (wantarray) {
        return @{$params->{$param}};
    }

    return $params->{$param}->[0];
}

sub params {
    my $self = shift;
    my %params;

    foreach my $param (@_) {
        my $value = $self->param($param);

        # If a value is not defined then the parameter does not exist
        # in the request. In this case the parameter is not set.
        if (defined $value) {
            $params{$param} = $self->param($param);
        }
    }

    return \%params;
}

sub exist {
    my ($self, $param) = @_;
    my $params = $self->{params};

    return exists $params->{$param} ? 1 : 0;
}

sub cookie {
    my $self = shift;

    if (@_ == 1) {
        my $name = shift;
        my $cookie = $self->http_cookie || "";

        if ($cookie) {
            foreach my $pair (split /\s*;\s*/, $cookie) {
                my ($key, $value) = split /\s*=\s*/, $pair;
                if ($key eq $name) {
                    return $value;
                }
            }
        }
        return "";
    }

    my %opts = @_;
    $opts{"-path"} //= "/";
    $opts{"-value"} //= "";

    if (!defined $opts{"-name"} || $opts{"-name"} !~ /\S/) {
        die "invalid cookie name: ". $opts{"-name"};
    }

    my @cookie = (join("=", $opts{"-name"}, $opts{"-value"}));

    if ($opts{"-expires"}) {
        push @cookie, "expires=" . $self->_expires($opts{"-expires"});
    }

    foreach my $key (qw/path domain/) {
        if (defined $opts{"-$key"}) {
            push @cookie, "$key=" . $opts{"-$key"};
        }
    }

    return join("; ", @cookie);
}

sub unescape {
    my ($self, $string) = @_;

    if (defined $string) {
        $string =~ s/(?:%([0-9A-Fa-f]{2})|\+)/defined $1 ? chr(hex($1)) : ' '/eg;
    }

    return $string;
}

sub upload {
    my $self = shift;

    return $self->{uploads} ? $self->{uploads}->[0] : undef;
}

sub uploads {
    my $self = shift;

    return wantarray ? @{$self->{uploads}} : $self->{uploads};
}

sub server_name          { $ENV{SERVER_NAME}          }
sub script_name          { $ENV{SCRIPT_NAME}          }
sub http_accept_encoding { $ENV{HTTP_ACCEPT_ENCODING} }
sub server_admin         { $ENV{SERVER_ADMIN}         }
sub http_connection      { $ENV{HTTP_CONNECTION}      }
sub http_accept          { $ENV{HTTP_ACCEPT}          }
sub request_method       { $ENV{REQUEST_METHOD}       }
sub script_filename      { $ENV{SCRIPT_FILENAME}      }
sub server_software      { $ENV{SERVER_SOFTWARE}      }
sub http_accept_charset  { $ENV{HTTP_ACCEPT_CHARSET}  }
sub http_cookie          { $ENV{HTTP_COOKIE}          }
sub http_te              { $ENV{HTTP_TE}              }
sub http_user_agent      { $ENV{HTTP_USER_AGENT}      }
sub remote_port          { $ENV{REMOTE_PORT}          }
sub query_string         { $ENV{QUERY_STRING}         }
sub server_signature     { $ENV{SERVER_SIGNATURE}     }
sub server_port          { $ENV{SERVER_PORT}          }
sub http_accept_language { $ENV{HTTP_ACCEPT_LANGUAGE} }
sub server_protocol      { $ENV{SERVER_PROTOCOL}      }
sub path                 { $ENV{PATH}                 }
sub path_info            { $ENV{SCRIPT_NAME}          }
sub gateway_interface    { $ENV{GATEWAY_INTERFACE}    }
sub request_uri          { $ENV{REQUEST_URI}          }
sub server_addr          { $ENV{SERVER_ADDR}          }
sub document_root        { $ENV{DOCUMENT_ROOT}        }
sub http_host            { $ENV{HTTP_HOST}            }
sub content_length       { $ENV{CONTENT_LENGTH} || 0  }

sub referer  { $ENV{HTTP_REFERRER} || $ENV{HTTP_REFERER} }
sub referrer { $ENV{HTTP_REFERRER} || $ENV{HTTP_REFERER} }

sub remote_addr {
    $ENV{HTTP_X_REAL_IP}
    || $ENV{X_REAL_IP}
    || $ENV{HTTP_X_FORWAREDED_FOR}
    || $ENV{X_FORWARDED_FOR}
    || $ENV{REMOTE_ADDR}
    || "::"
}

sub _process_query_string {
    my $self = shift;
    my $params = $self->{params};

    if ($_[0]) {
        foreach my $pair (split /[&;]/, $_[0]) {
            my ($param, $value) = split(/=/, $pair, 2);

            if (!defined $param) {
                next;
            }

            if (!defined $value) {
                $value = "";
            }

            $param = $self->unescape($param);
            $value = $self->unescape($value);
            $self->_add_param_value($param, $value);
        }
    }
}

sub _process_post_data {
    my $self = shift;
    my $params = $self->{params};

    if ($self->content_length > 0) {
        if ($self->{maxlen} && $self->content_length > $self->{maxlen}) {
            $self->log->warning("post data too long with", $self->content_length, "bytes");
        } else {
            read(\*STDIN, $self->{postdata}, $self->content_length);
        }
    }

    if ($self->{postdata}) {
        if ($self->content_type =~ m!^application/x-www-form-urlencoded!) {
            $self->_process_query_string($self->{postdata});
        } elsif ($self->content_type =~ m!^multipart/form-data!) {
            $self->_process_multipart($self->{postdata});
        } elsif ($self->is_json) {
            my $jsondata;
            local $SIG{__DIE__} = "DEFAULT";
            eval { $jsondata = JSON->new->utf8->decode($self->{postdata}) };

            if ($@) {
                $self->log->warning("unable to de-serialize json string - fixing string - $@");
                $self->{postdata} = Encode::decode("utf8", $self->{postdata});
                $self->{postdata} = Encode::encode("utf8", $self->{postdata});
                eval { $jsondata = JSON->new->utf8->decode($self->{postdata}) };
                if ($@) {
                    $self->log->warning("unable to de-serialize json string '$self->{postdata}' - $@");
                }
            } else {
                $self->{jsondata} = $jsondata;

                if (ref $jsondata eq "HASH") {
                    $self->{params} = $jsondata;
                }
            }
        }
    }
}

sub _process_multipart {
    my $self = shift;

    # ------WebKitFormBoundaryvPjZgS1pozQfuWN1
    # Content-Disposition: form-data; name="invoice_date"
    # 
    # 01.01.2012
    # ------WebKitFormBoundaryvPjZgS1pozQfuWN1
    # Content-Disposition: form-data; name="booking_date"
    # 
    # 01.01.2012
    # ------WebKitFormBoundaryvPjZgS1pozQfuWN1
    # Content-Disposition: form-data; name="value_date"
    # 
    # 01.01.2012
    # ------WebKitFormBoundaryvPjZgS1pozQfuWN1
    # Content-Disposition: form-data; name="purpose"
    # 
    # Foobar
    # ------WebKitFormBoundaryvPjZgS1pozQfuWN1
    # Content-Disposition: form-data; name="file"; filename=""
    # Content-Type: application/octet-stream
    # 
    # ------WebKitFormBoundaryvPjZgS1pozQfuWN1--

    my $crlf = CRLF;
    my $boundary = "--" . $self->boundary;
    my @parts = split /(?:^|$crlf)\Q$boundary\E/, $_[0];

    if ($parts[0] eq "") {
        shift @parts;
    }

    if ($parts[$#parts] =~ /^\-\-($crlf|\z)/) {
        pop @parts;
    }

    foreach my $part (@parts) {
        my ($filename, $name, $content_type, $charset);
        my ($header, $body) = $part =~ /^(?:$crlf)*(.+?)$crlf$crlf(.*)/s;
        $header //= "";
        $body //= "";

        foreach my $head (split /$crlf/, $header) {
            if ($head =~ /^Content\-Disposition:\s+form\-data;.*\sfilename=(["'])([^\1]*?)\1/) {
                $filename = $2;
                if ($head =~ /\sname=(["'])([^\1]*?)\1/) {
                    $name = $2;
                }
            } elsif ($head =~ /^Content\-Transfer\-Encoding:\s+base64/) {
                $body = MIME::Base64::decode_base64($body);
            } elsif ($head =~ m!^Content\-Type:\s+([a-zA-Z_0-9\-/]+)!) {
                $content_type = $1;
                if ($head =~ /\scharset=([^\s]+)/) {
                    $charset = $1;
                }
            } elsif ($head =~ /^Content\-Disposition:\s+form\-data;.*\sname=(["'])([^\1]*?)\1/) {
                $name = $2;
            }
        }

        if ($filename) {
            my $tmpfile = qx{/bin/mktemp --suffix ".DRAGON"};
            chomp $tmpfile;

            open my $fh, ">>", $tmpfile or die "unable to open tmpfile '$tmpfile' - $!";
            print $fh $body or die "unable to write to tmpfile '$tmpfile' - $!";
            close $fh or die "unable to close tmpfile - '$tmpfile' - $!";

            my $file = {
                size => length $body,
                filename => $filename,
                tmpfile => $tmpfile,
                charset => $charset // "",
                content_type => $content_type // ""
            };

            push @{$self->{uploads}}, $file;
            $body = $file;

        }

        if (defined $name && $name =~ /\S/) {
            $self->_add_param_value($name, $body);
        }
    }
}

sub _add_param_value {
    my ($self, $param, $value) = @_;
    my $params = $self->{params};

    if (exists $params->{$param}) {
        if (ref $params->{$param} ne "ARRAY") {
            $params->{$param} = [ $params->{$param} ];
        }
        push @{$params->{$param}}, $value;
    } else {
        $params->{$param} = $value;
    }
}

sub _expires {
    my ($self, $time) = @_;

    my @mon =qw(Jan Feb Mar Apr May Jun Jul Aug Sep Oct Nov Dec);
    my @wday = qw(Sun Mon Tue Wed Thu Fri Sat);
    my %t2s = (s => 1, m => 60, h => 3_600, d => 86_400, M => 2_592_000, y => 31_536_000);

    if ($time eq "now") {
        $time = time;
    } elsif ($time =~ /^([+-]?(?:\d+|\d*\.\d*))([smhdMy])/) {
        $time = time + ($t2s{$2} * $1);
    } elsif ($time !~ /^\d+/) {
        return $time;
    }

    my ($sec, $min, $hour, $mday, $mon, $year, $wday) = gmtime($time);
    $year += 1900;

    return sprintf(
        "%s, %02d-%s-%04d %02d:%02d:%02d GMT",
        $wday[$wday], $mday, $mon[$mon], $year, $hour, $min, $sec
    );
}

sub DESTROY {
    my $self = shift;

    if ($self->{uploads}) {
        foreach my $upload (@{$self->{uploads}}) {
            if ($upload->{filename} && -e $upload->{filename}) {
                unlink($upload->{filename});
            }
        }
    }
}

1;


lib/Bloonix/FCGI/SharedMem.pm
package Bloonix::FCGI::SharedMem;

use strict;
use warnings;
use Fcntl qw(O_CREAT O_RDWR :flock);
use IPC::SysV qw(IPC_PRIVATE S_IRWXU);
use IPC::SharedMem;

use constant PARENT_PID => $$;

# Process status:
#
# "_" Waiting for Connection
# "S" Starting up
# "R" Reading Request
# "W" Sending Reply
# "P" Processing
# "E" Error
# "C" Closing connection

sub new {
    my ($class, $slots, $lockfile) = @_;
    my $self = bless { }, $class;

    $self->{free_slots} = [ 0 .. $slots - 1 ];
    $self->{used_slots} = { };
    $self->{slot_size}  = 1024;
    $self->{slot_count} = $slots;

    if ($lockfile) {
        my $pid = PARENT_PID;
        $lockfile =~ s/%P/$pid/g;
    } else {
        $lockfile ||= join(".", "/var/run/blxipc", PARENT_PID, "tab");
    }

    if (-e $lockfile) {
        unlink($lockfile)
            or die "unable to remove lock file '$lockfile' - $!";
    }

    $self->{lockfile} = $lockfile;
    $self->_open_lockfile;

    $self->{shm} = IPC::SharedMem->new(
        IPC_PRIVATE,
        $slots * $self->{slot_size},
        S_IRWXU,
    );

    $self->{keys} = {
        # key - offset - size
        pid     => [   0,   5 ],
        status  => [   5,   1 ],
        time    => [   6,  10 ],
        ttlreq  => [  16,  20 ],
        client  => [  36,  39 ],
        request => [  75, 949 ],
    };

    return $self;
}

sub locking {
    my ($self, $bool) = @_;

    if (defined $bool) {
        $self->{locking} = $bool;
    }

    return $self->{locking};
}

sub get_free_slot {
    my $self = shift;
    my $slot = shift @{$self->{free_slots}};
    return $slot;
}

sub init_free_slot {
    my ($self, $slot, $pid, %pairs) = @_;
    $self->{used_slots}->{$pid} = $slot;

    $pairs{pid} //= $pid;
    $pairs{status} //= "S";
    $pairs{time} //= time;
    $pairs{ttlreq} //= 0;
    $pairs{client} //= "";
    $pairs{request} //= "initializing";

    $self->set($pid, %pairs);
}

sub wait_for_slot {
    my ($self, $slot, $pid) = @_;
    $self->{used_slots}->{$pid} = $slot;

    while ( 1 ) {
        my $p = $self->get($pid);

        if ($p->{pid} eq $pid) {
            last;
        }

        sleep 1;
    }
}

sub freeslots {
    my $self = shift;

    return scalar @{$self->{free_slots}};
}

sub remove {
    my ($self, $pid) = @_;
    my $slot = delete $self->{used_slots}->{$pid};
    my $length = $self->{slot_size};
    my $offset = ($slot * $length);
    my $value = "\0" x $length;

    $self->_lock;
    $self->_write($value, $offset, $length);
    $self->_unlock;

    push @{$self->{free_slots}}, $slot;
}

sub set {
    my ($self, $pid, %pairs) = @_;
    my $keys = $self->{keys};
    my $slot = $self->{used_slots}->{$pid};

    $self->_lock;

    while (my ($key, $value) = each %pairs) {
        die unless exists $keys->{$key};
        my ($offset, $length) = @{$keys->{$key}};
        $offset += ($slot * $self->{slot_size});
        $self->_write($value, $offset, $length);
    }

    $self->_unlock;
}

sub get {
    my ($self, $pid) = @_;
    my $used = $self->{used_slots};
    my $keys = $self->{keys};
    my $slot = $used->{$pid};
    my %pairs;

    if (!defined $slot) {
        return undef;
    }

    $self->_lock;

    foreach my $key (keys %$keys) {
        my ($offset, $length) = @{$keys->{$key}};
        $offset += ($slot * $self->{slot_size});
        $pairs{$key} = $self->_read($offset, $length);
    }

    $self->_unlock;
    return \%pairs;
}

sub proc_status {
    my $self = shift;
    my $keys = $self->{keys};
    my $fh = $self->{fh};
    my @proc_list;

    flock($fh, LOCK_EX);

    foreach my $slot (0..$self->{slot_count} - 1) {
        push @proc_list, \my %proc;
        foreach my $key (keys %$keys) {
            my ($offset, $length) = @{$keys->{$key}};
            $offset += ($slot * $self->{slot_size});

            my $value = $self->{shm}->read($offset, $length)
                or die "unable to read from shared memory - $!";

            $value =~ s/\0+\z//;
            $proc{$key} = $value;
        }
    }

    flock($fh, LOCK_UN);
    return \@proc_list;
}

sub _open_lockfile {
    my $self = shift;
    my $lockfile = $self->{lockfile};

    if ($self->{curpid} && $self->{curpid} == $$) {
        return $self->{fh};
    }

    sysopen my $fh, $lockfile, O_CREAT | O_RDWR
        or die "unable to create lockfile '$lockfile' - $!";

    $self->{fh} = $fh;
    $self->{curpid} = $$;

    return $fh;
}

sub _lock {
    my $self = shift;
    my $fh = $self->{fh};

    # Locking 0: LOCK_SH
    # Locking 1: LOCK_EX
    #
    # Why this dubious locking?
    #
    # The children will always write data in its slots and will
    # never overwrite data of other children. For this reason a
    # shared lock is adequate.
    #
    # The parent process needs to read a constistent state from the
    # shared memory of its children. For this reason the parent
    # process needs to lock always in exclusive mode for all
    # operations.
    #
    # In the end a child can only obtain a shared lock if the parent
    # does not lock in exlusive mode.

    if ($self->{locking}) {
        flock($fh, LOCK_EX);
    } else {
        flock($fh, LOCK_SH);
    }
}

sub _unlock {
    my $self = shift;
    my $fh = $self->{fh};
    flock($fh, LOCK_UN);
}

sub _read {
    my ($self, $offset, $length) = @_;

    my $string = $self->{shm}->read($offset, $length)
        or die "unable to read from shared memory - $!";

    $string =~ s/\0+\z//;
    return $string;
}

sub _write {
    my ($self, $string, $offset, $length) = @_;

    if (!defined $string) {
        $string = "";
    }

    if (length $string > $length) {
        $string = substr($string, 0, $length);
    } elsif (length $string < $length) {
        $string .= "\0" x ($length - length $string);
    }

    $self->{shm}->write($string, $offset, $length)
        or die "unable to write to shared memory - $!";
}

sub destroy {
    my $self = shift;
    $self->DESTROY;
}

sub DESTROY {
    if ($$ == PARENT_PID) {
        my $self = shift;
        if ($self->{shm}) {
            $self->{shm}->remove;
        }
    }
}

1;


lib/Bloonix/FCGI/SharedFile.pm
package Bloonix::FCGI::SharedFile;

use strict;
use warnings;
use Fcntl qw( O_CREAT O_RDWR :flock );

use constant PARENT_PID => $$;
use constant RECVBUF => 16384;

sub new {
    my ($class, $slots, $lockfile) = @_;
    my $self = bless { }, $class;

    $self->{free_slots} = [ 0 .. $slots - 1 ];
    $self->{used_slots} = { };
    $self->{slot_size}  = 1024;
    $self->{slot_count} = $slots;

    if ($lockfile) {
        my $pid = PARENT_PID;
        $lockfile =~ s/%P/$pid/g;
    } else {
        $lockfile ||= join(".", "/var/run/blxipc", PARENT_PID, "tab");
    }

    if (-e $lockfile) {
        unlink($lockfile)
            or die "unable to remove lock file '$lockfile' - $!";
    }

    $self->{lockfile} = $lockfile;
    my $fh = $self->_open_lockfile;
    print $fh "\0" x ($slots * $self->{slot_size});

    $self->{keys} = {
        # key - offset - size
        pid     => [   0,   5 ],
        status  => [   5,   1 ],
        time    => [   6,  10 ],
        ttlreq  => [  16,  20 ],
        client  => [  36,  39 ],
        request => [  75, 949 ],
    };

    return $self;
}

sub locking {
    my ($self, $bool) = @_;

    if (defined $bool) {
        $self->{locking} = $bool;
    }

    return $self->{locking};
}

sub get_free_slot {
    my $self = shift;
    my $slot = shift @{$self->{free_slots}};
    return $slot;
}

sub init_free_slot {
    my ($self, $slot, $pid, %pairs) = @_;
    $self->{used_slots}->{$pid} = $slot;

    $pairs{pid} //= $pid;
    $pairs{status} //= "S";
    $pairs{time} //= time;
    $pairs{ttlreq} //= 0;
    $pairs{client} //= "0.0.0.0";
    $pairs{request} //= "initializing";

    $self->set($pid, %pairs);
}

sub wait_for_slot {
    my ($self, $slot, $pid) = @_;
    $self->{used_slots}->{$pid} = $slot;

    while ( 1 ) {
        my $p = $self->get($pid);

        if ($p->{pid} eq $pid) {
            last;
        }

        sleep 1;
    }
}

sub freeslots {
    my $self = shift;

    return scalar @{$self->{free_slots}};
}

sub remove {
    my ($self, $pid) = @_;
    my $slot = delete $self->{used_slots}->{$pid};
    my $length = $self->{slot_size};
    my $offset = ($slot * $length);
    my $value = "\0" x $length;

    $self->_lock;
    $self->_write($value, $offset, $length);
    $self->_unlock;

    push @{$self->{free_slots}}, $slot;
}

sub set {
    my ($self, $pid, %pairs) = @_;
    my $keys = $self->{keys};
    my $slot = $self->{used_slots}->{$pid};

    $self->_lock;

    foreach my $key (keys %pairs) {
        die unless exists $keys->{$key};
        my ($offset, $length) = @{$keys->{$key}};
        $offset += ($slot * $self->{slot_size});
        $self->_write($pairs{$key}, $offset, $length);
    }

    $self->_unlock;
}

sub get {
    my ($self, $pid) = @_;
    my $used = $self->{used_slots};
    my $keys = $self->{keys};
    my $slot = $used->{$pid};
    my %pairs;

    if (!defined $slot) {
        return undef;
    }

    $self->_lock;

    foreach my $key (keys %$keys) {
        my ($offset, $length) = @{$keys->{$key}};
        $offset += ($slot * $self->{slot_size});
        $pairs{$key} = $self->_read($offset, $length);
    }

    $self->_unlock;
    return \%pairs;
}

sub proc_status {
    my $self = shift;
    my $keys = $self->{keys};
    my $fh = $self->{fh};
    my @proc_list;

    flock($fh, LOCK_EX);

    foreach my $slot (0..$self->{slot_count} - 1) {
        push @proc_list, \my %proc;
        foreach my $key (keys %$keys) {
            my ($offset, $length) = @{$keys->{$key}};
            $offset += ($slot * $self->{slot_size});
            $proc{$key} = $self->_read($offset, $length);
        }
    }

    flock($fh, LOCK_UN);
    return \@proc_list;
}

sub _open_lockfile {
    my $self = shift;
    my $lockfile = $self->{lockfile};

    if ($self->{curpid} && $self->{curpid} == $$) {
        return $self->{fh};
    }

    sysopen my $fh, $lockfile, O_CREAT | O_RDWR
        or die "unable to create lockfile '$lockfile' - $!";

    $self->{fh} = $fh;
    $self->{curpid} = $$;

    my $oldfh = select $fh;
    $| = 1;
    select $oldfh;
    return $fh;
}

sub _lock {
    my $self = shift;
    my $fh = $self->{fh};

    # Locking 0: LOCK_SH
    # Locking 1: LOCK_EX
    #
    # Why this dubious locking?
    #
    # The children will always write data in its slots and will
    # never overwrite data of other children. For this reason a
    # shared lock is adequate.
    #
    # The parent process needs to read a constistent state of the
    # from the shared memory of its children. For this reason the
    # parent process needs to lock always in exclusive mode for all
    # operations.
    #
    # In the end a child can only obtain a shared lock if the parent
    # does not lock in exlusive mode.

    if ($self->{locking}) {
        flock($fh, LOCK_EX);
    } else {
        flock($fh, LOCK_SH);
    }
}

sub _unlock {
    my $self = shift;
    my $fh = $self->{fh};
    flock($fh, LOCK_UN);
}

sub _write {
    my ($self, $string, $offset, $length) = @_;
    my $fh = $self->_open_lockfile;
    my $rest = $length;
    my $written = 0;

    if (!defined $string) {
        $string = "";
    }

    if (length $string > $length) {
        $string = substr($string, 0, $length);
    } elsif (length $string < $length) {
        $string .= "\0" x ($length - length $string);
    }

    $self->_seek($fh, $offset);
    $offset = 0;

    while ($rest) {
        $written = syswrite $fh, $string, $rest, $offset;

        if (!defined $written) {
            die "unable to write to shared file - $!";
        } elsif ($written) {
            $rest -= $written;
            $offset += $written;
        }
    }
}

sub _seek {
    my ($self, $fh, $offset) = @_;
    my $counter = 0;

    while ($offset != sysseek($fh, 0, 1)) {
        sysseek($fh, $offset, 0);
        $counter++;

        if ($counter > 10) {
            warn "ERR: unable to seek to lock file pos $offset";
            sleep 1;
        }
    }
}

sub _read {
    my ($self, $offset, $length) = @_;
    my ($data, $len, $buf);
    my $fh = $self->_open_lockfile;
    my $rest = $length;
    my $rdsz = $length < RECVBUF ? $length : RECVBUF;
    my $clen = 0;

    $self->_seek($fh, $offset);

    while ($rest) {
        $len = sysread $fh, $buf, $rdsz;

        if (!defined $len) {
            if ($! =~ /^Interrupted/) {
                Time::HiRes::sleep(0.1); # careful
                next;
            }
            die "unable to read from shared file - $!";
        } elsif ($len) {
            $data .= $buf;  # concat the data pieces
            $rest -= $len;  # this is the rest we have to read
            #$clen += $len;  # the current len
            #warn "read $clen/$length bytes";
        } elsif ($rest) {
            # Oh, doh, this should never happends.
            # Maybe the other end dies.
            #die "unable to read all data from client (rest $rest, length $len, read size $rdsz)";
            Time::HiRes::sleep(0.1); # careful
        }

        if ($rest < $rdsz) {
            # otherwise sysread() hangs if we wants to read to much
            $rdsz = $rest;
        }
    }

    $data =~ s/\0+\z//;
    return $data;
}

sub destroy {
    my $self = shift;
    $self->DESTROY;
}

sub DESTROY {
    if ($$ == PARENT_PID) {
        my $self = shift;
        unlink($self->{lockfile});
    }
}

1;
